lcls_tools/common/data/emittance/fit.py

```diff
@@ -0,0 +1,189 @@
+
+
+def compute_emit_bmag(
+    k: np.ndarray,
+    beamsize_squared: np.ndarray,
+    q_len: float,
+    rmat: np.ndarray,
+    twiss_design: np.ndarray = None,
+    thin_lens: bool = False,
+    maxiter: int = None,
+):
+    """
+    Computes the emittance(s) corresponding to a set of quadrupole measurement scans
+    using nonlinear fitting of beam matrix parameters to guarantee physically valid results.
+
+    Parameters:
+        k: numpy array of shape (n_steps_quad_scan,) or (batchshape x n_steps_quad_scan)
+            representing the measurement quad geometric focusing strengths in [m^-2]
+            used in the emittance scan(s)
+
+        beamsize_squared: numpy array of shape (batchshape x n_steps_quad_scan),
+                representing the mean-square beamsize outputs in [mm^2] of the emittance scan(s)
+                with inputs given by k
+
+        q_len: float defining the (longitudinal) quadrupole length or "thickness" in [m]
+
+        rmat: numpy array shape (2x2) or (batchshape x 2 x 2)
+                containing the 2x2 R matrices describing the transport from the end of the
+                measurement quad to the observation screen.
+
+        twiss_design: numpy array shape (2,) or (batchshape x 2) designating the design (beta, alpha) 
+                        twiss parameters at the screen.
+
+        thin_lens: boolean specifying whether or not to use thin lens approximation for measurement quad.
+
+        max_iter: maximum number of iterations to perform in nonlinear fitting (minimization algorithm)
+
+    Returns:
+        emit: numpy array shape (batchshape x 1) containing the geometric emittance fit results for each scan in mm-mrad
+        bmag: numpy array shape (batchshape x n_steps) containing the bmag corresponding to each point in each scan
+        beam_matrix: numpy array shape (batchshape x 3) containing [sig11, sig12, sig22] where sig11, sig12, sig22
+                        are the reconstructed beam matrix parameters at the entrance of the measurement quad
+        twiss_at_screen: numpy array shape (batchshape x nsteps x 3) containing the reconstructed twiss parameters
+                            at the measurement screen for each step in each quad scan.
+
+    SOURCE PAPER: http://www-library.desy.de/preparch/desy/thesis/desy-thesis-05-014.pdf
+    """
+    # construct the A matrix from eq. (3.2) & (3.3) of source paper
+    quad_rmat = build_quad_rmat(
+        k, q_len, thin_lens=thin_lens
+    )  # result shape (batchshape x nsteps x 2 x 2)
+    total_rmat = np.expand_dims(rmat, -3) @ quad_rmat
+    # result shape (batchshape x nsteps x 2 x 2)
+
+    # prepare the A matrix
+    r11, r12 = total_rmat[..., 0, 0], total_rmat[..., 0, 1]
+    amat = np.stack((r11**2, 2.0 * r11 * r12, r12**2), axis=-1)
+    # amat result (batchshape x nsteps x 3)
+
+    beamsize_squared = np.expand_dims(beamsize_squared, -1)
+
+    def beam_matrix_tuple(params):
+        '''
```

Comment:
Is it a requirement that this is inside the bmag function? Can we pull this out?

Reply:
It is not a requirement, however it is a function whose only use-case is within the function where I have defined it. Would you prefer that I move it outside of that function?

Reply:
ok, so there's no future case where we might want to use this somewhere else?

Reply:
Correct, it is just a convenience function for translating between fit parameters and the real beam matrix parameters, which is only useful during the fitting. I cannot conceive of a situation in which we would be interested in the transformed parameters outside of the fitting process.